### DISCUSSION TITLE
url_for support in templates?

### DISCUSSION THREAD

**Discussion by itcld** [mitsuhiko/minijinja#314] [Q&A] [answered]
Anyone know how to get `url_for` working in minijinja to generate links to the app, using Actix Web?

Actually I'm new to Rust itself (coming from C# and Python) and haven't been able to figure this out for any of the other Rust templating engines either (Tera, Askama).  Actix Web does have a [url_for on HttpRequest](https://docs.rs/actix-web/latest/actix_web/struct.HttpRequest.html#method.url_for), but not sure what I need to do to integrate it with minijinja, or if someone has done it.

**Comment by mitsuhiko** [comment, marked as answer]
`actix-web` makes this incredibly easy because the request object is cheaply clonable and can be stashed into a thread local. This means you can put it there before rendering, and then fetch that object back in a `url_for` function you can add to the template. I added an example that shows this: https://github.com/mitsuhiko/minijinja/blob/main/examples/actix-web-demo/src/main.rs

**Reply by itcld** [reply]
Thanks, that seems to work, at least in my limited testing.  I did tweak your `url_for` implementation to use [from_safe_string](https://docs.rs/minijinja/latest/minijinja/value/struct.Value.html#method.from_safe_string), so I could get links that get rendered like this:
`<a href="http://localhost:8080/users">Users</a>`
Instead of this:
`<a href="http:&#x2f;&#x2f;localhost:8080&#x2f;users">Users</a>`

I still need to figure out how to remove the `http://localhost:8080` part and just get the relative URL, but I think that is an issue with Actix Web `HttpRequest::url_for` always returning absolute URL.

**Reply by mitsuhiko** [reply]
@itcld `from_safe_string` is wrong here. You want your link to look like `http:&#x2f;&#x2f;localhost:8080&#x2f;users`. It's to prevent XSS in certain contexts.

**Reply by step21** [reply]
would sth similar work for axum as well?

**Reply by mitsuhiko** [reply]
@step21 does axum have an API to generate URLs?

**Reply by step21** [reply]
uhm, probably, but I am still learning it, so not sure. Maybe this? https://docs.rs/axum/latest/axum/extract/struct.Query.html#impl-Clone-for-Query%3CT%3E

**Reply by mitsuhiko** [reply]
I was unable to find an API that generates routes. This issue implies it's not possible: https://github.com/tokio-rs/axum/discussions/1947

**Reply by step21** [reply]
thanks. do you maybe have an alternative idea for just making sure that links have the correct domain etc? F.e. in some of my cases I mainly use it so that files are found both in dev and in prod.
F.e. 
```
<script src="{{ url_for('static', filename='static/js/string.min.js') }}"></script>
```

**Reply by mitsuhiko** [reply]
I'm not sure if axum has that information either.

**Comment by surajsharma** [comment]
I'm using rocket but cannot find something equivalent, any hints?

**Reply by mitsuhiko** [reply]
I think you should ask this question in the rocket repo.  I'm not sure if rocket supports dynamic route generation as the documented way ([uri!](https://docs.rs/rocket/latest/rocket/macro.uri.html)) involves code generation.

**Reply by surajsharma** [reply]
@mitsuhiko using static routes for now, thanks for the quick response!
